test/Microsoft.ML.Functional.Tests/Explainability.cs

```diff
@@ -0,0 +1,284 @@
+
+            // Create a pipeline to train on the housing data.
+            var pipeline = mlContext.Transforms.Concatenate("Features", HousingRegression.Features)
+                .Append(mlContext.Regression.Trainers.FastTree());
+
+            // Fit the pipeline and transform the data.
+            var model = pipeline.Fit(data);
+            var treeModel = model.LastTransformer.Model;
+
+            // Get the feature gain.
+            var weights = new VBuffer<float>();
+            treeModel.GetFeatureWeights(ref weights);
+
+            // Make sure the number of feature gains returned matches the length of the input feature vector.
+            Assert.Equal(HousingRegression.Features.Length, weights.Length);
+        }
+
+        /// <summary>
+        /// GlobalFeatureImportance: A FastForest model can give back global feature importance through feature gain.
+        /// </summary>
+        [Fact]
+        public void GlobalFeatureImportanceForFastForestThroughFeatureGain()
+        {
+            var mlContext = new MLContext(seed: 1, conc: 1);
+
+            // Get the dataset
+            var data = mlContext.Data.ReadFromTextFile<HousingRegression>(GetDataPath(TestDatasets.housing.trainFilename), hasHeader: true);
+
+            // Create a pipeline to train on the housing data.
+            var pipeline = mlContext.Transforms.Concatenate("Features", HousingRegression.Features)
+                .Append(mlContext.Regression.Trainers.FastForest());
+
+            // Fit the pipeline and transform the data.
+            var model = pipeline.Fit(data);
+            var treeModel = model.LastTransformer.Model;
+
+            // Get the feature gain
+            var weights = new VBuffer<float>();
+            treeModel.GetFeatureWeights(ref weights);
+
+            // Make sure the number of feature gains returned matches the length of the input feature vector.
+            Assert.Equal(HousingRegression.Features.Length, weights.Length);
+        }
+
+        /// <summary>
+        /// LocalFeatureImportance: Per-row feature importance can be computed through FeatureContributionCalculator for a linear model.
+        /// </summary>
+        [Fact]
+        public void LocalFeatureImportanceForLinearModel()
+        {
+            var mlContext = new MLContext(seed: 1, conc: 1);
+
+            // Get the dataset
+            var data = mlContext.Data.ReadFromTextFile<HousingRegression>(GetDataPath(TestDatasets.housing.trainFilename), hasHeader: true);
+
+            // Create a pipeline to train on the housing data.
+            var pipeline = mlContext.Transforms.Concatenate("Features", HousingRegression.Features)
+                .Append(mlContext.Regression.Trainers.StochasticDualCoordinateAscent());
+
+            // Fit the pipeline and transform the data.
+            var model = pipeline.Fit(data);
+            var scoredData = model.Transform(data);
```

Comment:
this seems common to all tests. factor out.

Reply:
I tried to do this, but unfortunately we no longer have a way to specify a generic model that can be used with the `FeatureContributionCalculator`, as we made all the interfaces internal.

---
In reply to: [258591290](https://github.com/dotnet/machinelearning/pull/2584#discussion_r258591290) [](ancestors = 258591290)

Reply:
bummer

---
In reply to: [259094300](https://github.com/dotnet/machinelearning/pull/2584#discussion_r259094300) [](ancestors = 259094300,258591290)